<a href="https://colab.research.google.com/github/notfirsttt/Dasar_Kecerdasan_Artifisial/blob/main/PertemuanMinggu2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Guided 1

In [ ]:
import numpy as np
x = np.array([4, 9, 2, 5])
y = np.array([3, 5, 3, 2])

k = np.concatenate((x, y))
z = x + y

print(k)
print(z)
print(z.tolist())

[4 9 2 5 3 5 3 2]
[ 7 14  5  7]
[7, 14, 5, 7]


Kode di atas make NumPy buat ngolah dua array 1D, yaitu x dan y. Pertama, x diisi dengan [4, 9, 2, 5] dan y diisi dengan [3, 5, 3, 2]. Terus, np.concatenate((x, y)) dipake buat ngegabungin kedua array tanpa operasi matematika, jadi hasilnya [4, 9, 2, 5, 3, 5, 3, 2]. Setelah itu, x + y ngejumlahin elemen yang ada di indeks yang sama, jadinya [7, 14, 5, 7]. Terakhir, z.tolist() ngeubah hasil penjumlahan dari array NumPy jadi list biasa di Python, dengan output yang sama, yaitu [7, 14, 5, 7].

##Guided 2


In [ ]:
import numpy as np

x = np.array([[5, 3, 2, 1], [6, 3, 2, 1]])
print("Dimensi aray x: ", x.ndim)

Dimensi aray x:  2


Kode di atas make NumPy buat bikin array 2D yang isinya dua baris, yaitu [5, 3, 2, 1] dan [6, 3, 2, 1]. Terus, fungsi ndim dipake buat ngecek jumlah dimensi dari array x. Karena x punya dua baris dan beberapa kolom, berarti array ini termasuk array 2D, jadi output dari x.ndim adalah 2. Fitur ini berguna banget buat ngecek struktur data sebelum ngelakuin operasi lebih lanjut, biar gak salah paham sama bentuk array yang diproses.

##Guided3

In [ ]:
import numpy as np

x = np.array([[5, 3, 2, 1], [6, 3, 2, 1]])
y = np.array([5.2, 3.3, 8.1])
print("Shape x  : ", x.shape)
print("Size x   : ", x.size)
print("Shape y  : ", y.dtype)
print("Size y   : ", y.dtype)


Shape x  :  (2, 4)
Size x   :  8
Shape y  :  float64
Size y   :  float64


Kode di atas make NumPy buat bikin dua array, yaitu x yang berbentuk 2D dengan dua baris dan empat kolom, serta y yang berbentuk 1D dengan tiga elemen. Fungsi shape dipake buat ngecek ukuran atau dimensi array, jadi x.shape bakal ngeluarin (2, 4) yang nunjukin dua baris dan empat kolom. Terus, fungsi size dipake buat ngitung total elemen dalam array, jadi x.size bakal hasilin 8 karena ada delapan angka di dalamnya. Sedangkan untuk array y, fungsi dtype dipake buat ngecek tipe data elemen di dalamnya, yang dalam hal ini adalah float64, karena isinya angka desimal.

##Guided 4

In [ ]:
import numpy as np


x = np.zeros((3,3))
y = np.ones((2,3))
z = np.empty((2,2))
k = np.arange(0, 10, 2)
l = np.linspace(0, 1, 5)
m = np.full((2, 2), 5)
n = np.random.random((2, 3))
o = np.random.randint(0, 50, size = (2,3))


print(x)
print(y)
print(z)
print(k)
print(l)
print(m)
print(n)
print(o)

[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]
[[1. 1. 1.]
 [1. 1. 1.]]
[[0.25 0.5 ]
 [0.75 1.  ]]
[0 2 4 6 8]
[0.   0.25 0.5  0.75 1.  ]
[[5 5]
 [5 5]]
[[0.77224051 0.96271273 0.00473575]
 [0.59235707 0.08086049 0.52890981]]
[[24  6 25]
 [28 22 49]]


Kode di atas make NumPy buat bikin berbagai jenis array dengan metode yang beda-beda. Pertama, np.zeros((3,3)) dipake buat bikin array 3x3 yang isinya semua nol. Terus, np.ones((2,3)) bikin array 2x3 yang semuanya berisi angka satu. np.empty((2,2)) bikin array 2x2 dengan nilai acak yang bisa berubah tergantung kondisi memori. np.arange(0, 10, 2) bikin array 1D dengan angka dari 0 sampai 10 tapi lompat 2, jadi hasilnya 0, 2, 4, 6, 8. np.linspace(0, 1, 5) bikin lima angka yang dibagi rata antara 0 sampai 1, hasilnya 0, 0.25, 0.5, 0.75, 1. np.full((2,2), 5) bikin array 2x2 yang semuanya angka 5. np.random.random((2,3)) bikin array 2x3 dengan angka random antara 0 sampai 1, sedangkan np.random.randint(0, 50, size=(2,3)) bikin array 2x3 dengan angka random dari 0 sampai 49. Semua hasilnya bakal ditampilin di output sesuai jenis array yang dibuat

##Guided 5

In [ ]:
import numpy as np
#append pada array 1D
a = np.array([[1, 2, 3]])
a = np.append(a, [4, 5])
print("1D array setelah append:\n", a)
#append pada array 2D
x = np.array([[1, 2], [3, 4]])
x = np.append(x, [4, 5, 6])
print("2D array setelah append:\n", x)

1D array setelah append:
 [1 2 3 4 5]
2D array setelah append:
 [1 2 3 4 4 5 6]


Kode di atas make NumPy buat nambahin elemen ke array 1D dan 2D pake fungsi np.append. Pertama, array a yang awalnya [[1, 2, 3]] ditambahin elemen [4, 5], hasilnya jadi [1, 2, 3, 4, 5]. Fungsi np.append nambahin elemen di akhir array tanpa ngubah dimensi awalnya. Terus, array x yang awalnya [[1, 2], [3, 4]] ditambahin [4, 5, 6], tapi karena jumlah elemen yang ditambahin ga sesuai sama dimensi awal, hasilnya jadi array 1D [1, 2, 3, 4, 4, 5, 6]. Kalo mau tetap di bentuk 2D, sebaiknya nambahinnya dalam bentuk baris atau kolom yang sesuai dengan struktur array aslinya

##Guided 6

In [ ]:
import numpy as np

# concatenate pada array 1D
a = np.array([1, 2])
b = np.array([3, 4])
c = np.concatenate((a, b))
print("1D array setelah concatenate:\n", c)
# concatenate pada array 2D
x = np.array([[1, 2], [3, 4]])
y = np.array([[5, 6]])
z = np.concatenate((x, y), axis=0)
print("1D array setelah concatenate:\n", z)

1D array setelah concatenate:
 [1 2 3 4]
1D array setelah concatenate:
 [[1 2]
 [3 4]
 [5 6]]


Kode di atas make NumPy buat ngegabungin dua array pake fungsi np.concatenate. Pertama, ada array a yang isinya [1, 2] dan b yang isinya [3, 4]. Pas digabungin pake np.concatenate((a, b)), hasilnya jadi [1, 2, 3, 4], yang berarti elemen dari b ditambahin langsung ke belakang a. Selanjutnya, ada array 2D x yang isinya [[1, 2], [3, 4]] dan y yang isinya [[5, 6]]. Pas digabungin pake np.concatenate((x, y), axis=0), hasilnya jadi [[1, 2], [3, 4], [5, 6]], yang artinya baris dari y ditambahin ke bawah x. Kalo mau gabungin berdasarkan kolom, tinggal ubah axis=1 supaya elemen dari y ditambahkan sebagai kolom tambahan di x

##Guided 7

In [ ]:
import numpy as np

# insert pada array 1D
a = np.array([1, 2, 3])
a = np.insert(a, 1, 4) # Insert '4' pada index 1
print("1D array setelah insert:\n", a)
# insert pada array 2D
x = np.array([[1, 2], [3, 4]])
x = np.insert(x, 0, [5, 6], axis=0) # Insert baris [5, 6] index 0
print("2D array setelah insert:\n", x)

1D array setelah insert:
 [1 4 2 3]
2D array setelah insert:
 [[5 6]
 [1 2]
 [3 4]]


Kode di atas make fungsi np.insert dari NumPy buat nyisipin elemen ke dalam array 1D dan 2D. Di bagian pertama, ada array 1D [1, 2, 3], terus elemen 4 dimasukin ke indeks 1, jadi hasil akhirnya berubah jadi [1, 4, 2, 3]. Fungsi np.insert(a, 1, 4) ini nggak ngeganti elemen yang udah ada, tapi langsung nyelipin angka 4 di posisi yang ditentuin.  
Lanjut ke array 2D [[1, 2], [3, 4]], baris [5, 6] ditambahin di indeks 0 pake np.insert(x, 0, [5, 6], axis=0). Karena axis=0, elemen baru ini disisipin sebagai baris, jadi hasil akhirnya jadi [[5, 6], [1, 2], [3, 4]]. Secara umum, np.insert bisa dipake buat nyisipin elemen ke indeks tertentu dalam array, baik itu di baris maupun kolom, tergantung dari nilai axis-nya. Kalau axis=None, array bakal diratain dulu jadi satu dimensi sebelum elemen baru disisipin.

##Guided 8

In [ ]:
import numpy as np

x = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
x = np.delete(x, 1, axis=0)
print(x)

[[1 2 3]
 [7 8 9]]


Kode di atas make NumPy buat ngehapus baris di array 2D pake fungsi np.delete. Pertama, ada array x yang bentuknya matriks 3x3 dengan elemen [[1, 2, 3], [4, 5, 6], [7, 8, 9]]. Terus, np.delete(x, 1, axis=0) dipake buat ngapus baris dengan indeks 1, yaitu baris kedua [4, 5, 6]. Hasil akhirnya jadi [[1, 2, 3], [7, 8, 9]], di mana baris yang dihapus udah gak ada lagi di array

##Guided 9

In [ ]:
import numpy as np

x = np.array([4, 2, 7])
y = np.sort(x)
print("Array yang telah diurutkan:", y)

Array yang telah diurutkan: [2 4 7]


Kode di atas make NumPy buat ngurutkan elemen dalam array 1D pake fungsi np.sort. Pertama, ada array x yang isinya [4, 2, 7]. Terus, np.sort(x) dipake buat ngurutkan elemen dalam urutan naik, jadi hasilnya [2, 4, 7]. Hasil urutan ini disimpen di variabel y, dan pas dicetak, bakal nampilin array yang udah terurut

##Guided 10

In [ ]:
import numpy as np

x = np.array([[3, 2, 4], [1, 4, 5]])
y = np.sort(x, axis=0) # Mengurutkan sepanjang kolom
print("Array yang telah diurutkan sepanjang kolom:\n", y)

z = np.sort(x) # Mengurutkan sepanjang baris
print("Array yang telah diurutkan sepanjang baris:\n", z)

Array yang telah diurutkan sepanjang kolom:
 [[1 2 4]
 [3 4 5]]
Array yang telah diurutkan sepanjang baris:
 [[2 3 4]
 [1 4 5]]


Kode di atas make NumPy buat ngurutkan elemen dalam array 2D baik berdasarkan kolom maupun baris. Pertama, ada array x dengan nilai [[3, 2, 4], [1, 4, 5]]. Terus, np.sort(x, axis=0) dipake buat ngurutkan elemen berdasarkan kolom, jadi hasilnya [[1, 2, 4], [3, 4, 5]], yang disimpen di variabel y. Selanjutnya, np.sort(x) atau np.sort(x, axis=1) dipake buat ngurutkan elemen berdasarkan baris, jadi tiap barisnya bakal diurutkan sendiri, hasilnya [[2, 3, 4], [1, 4, 5]], yang disimpen di variabel z. Kedua hasil sorting ini kemudian dicetak

##Guided 11

In [ ]:
import numpy as np

x = np.array([20, 80, 10, 50, 30])
subset = x[1:4]

#indexing
print(x[2])

print(subset)

10
[80 10 50]


Kode di atas make NumPy buat ngeakses elemen dalam array pakai indexing dan slicing. Array x diinisialisasi dengan nilai [20, 80, 10, 50, 30]. Terus, subset = x[1:4] dipake buat nyimpen elemen dari indeks 1 sampai 3, hasilnya [80, 10, 50]. Buat indexing langsung, x[2] ngeambil elemen di indeks ke-2, yaitu 10. Terakhir, hasil indexing dan subset dicetak.

##Guided 12

In [ ]:
import numpy as np

x = np.array([3, 2, 6, 5])
# Membuat shallow copy menggunakan view()
y = x.view()
# Mengubah elemen pertama dari shallow copy
y[0] = 99
print("Array asli:", x)
print("Shallow copy:", y)

Array asli: [99  2  6  5]
Shallow copy: [99  2  6  5]


Kode di atas make NumPy buat bikin shallow copy dari array pakai fungsi view(). Awalnya, array x diinisialisasi dengan nilai [3, 2, 6, 5]. Terus, y = x.view() dipake buat bikin shallow copy, yang artinya y bakal nge-reference data yang sama dengan x tanpa ngegandain memori. Pas elemen pertama di y diubah jadi 99, perubahan ini juga ngaruh ke x karena keduanya masih terhubung ke data yang sama. Hasil akhirnya, array asli x berubah jadi [99, 2, 6, 5], dan shallow copy y juga berisi [99, 2, 6, 5].

##Guided 13

In [ ]:
import numpy as np

x = np.array([3, 2, 6, 5])

y = x.copy()
y[0] = 99

print("Array asli:", x)
print("Deep copy:", y)

Array asli: [3 2 6 5]
Deep copy: [99  2  6  5]


Kode di atas make NumPy buat bikin deep copy dari array pakai fungsi copy(). Awalnya, array x diinisialisasi dengan nilai [3, 2, 6, 5]. Terus, y = x.copy() dipake buat bikin deep copy, yang artinya y bakal jadi duplikasi penuh dari x tanpa berbagi memori. Pas elemen pertama di y diubah jadi 99, perubahan ini gak ngaruh ke x karena mereka udah terpisah. Hasil akhirnya, array asli x tetap [3, 2, 6, 5], sementara deep copy y berubah jadi [99, 2, 6, 5].

##Guided 14

In [ ]:
import numpy as np

x = np.array ([7, 2, 9, 3])

#Nilai Maksimal
max_value = np.max(x)
print("Nilai maks : ", max_value)

#Nilai Minimum
min_value = np.min(x)
print("Nilai min : ", min_value)

#Penjumlahan
sum_value = np.sum(x)
print("Jumlah : ", sum_value)

#Rata-rata
mean_value = np.mean(x)
print("Rata-rata : ", mean_value)

Nilai maks :  9
Nilai min :  2
Jumlah :  21
Rata-rata :  5.25


Kode di atas make NumPy buat ngolah array x yang berisi [7, 2, 9, 3]. Pertama, np.max(x) dipake buat nyari nilai maksimal dalam array, hasilnya 9. Terus, np.min(x) dipake buat nyari nilai minimum, hasilnya 2. Setelah itu, np.sum(x) ngejumlahin semua elemen dalam array, hasilnya 21. Terakhir, np.mean(x) dipake buat ngitung rata-rata elemen dalam array, hasilnya 5.25.

##Unguided 1

##Unguided 2